dhall-docs/src/Dhall/Docs/CodeRenderer.hs

```diff
@@ -0,0 +1,149 @@
+{-# LANGUAGE NamedFieldPuns    #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+
+{-| Contains the logic to render the source code inside a HTML. It also provides
+    context-sensitive features such as jump-to-definition
+-}
+module Dhall.Docs.CodeRenderer
+    ( exprSrcToHtml
+    , exprVoidToHtml
+    , ExprType(..)
+    ) where
+
+import Data.Text       (Text)
+import Data.Void       (Void)
+import Dhall.Core
+    ( Expr (..)
+    , File (..)
+    , FilePrefix (..)
+    , Import (..)
+    , ImportHashed (..)
+    , ImportType (..)
+    , Scheme (..)
+    , URL (..)
+    )
+import Dhall.Docs.Util
+import Dhall.Src       (Src (..))
+import Lucid
+
+import qualified Data.Text
+import qualified Data.Text.Prettyprint.Doc             as Pretty
+import qualified Data.Text.Prettyprint.Doc.Render.Text as Pretty.Text
+import qualified Dhall.Core                            as Core
+import qualified Dhall.Parser
+import qualified Dhall.Pretty
+import qualified Lens.Family                           as Lens
+import qualified Text.Megaparsec.Pos                   as SourcePos
+
+-- | Given a Dhall expression return all imports with their location on the file.
+--   Contents are already sorted by 'Src', which allows 'renderAsHtml' to traverse
+--   the return of this function to inject anchors
+getImports :: Expr Src Import -> [(Src, Import)]
+getImports (Note src (Embed a)) = [(src, a)]
+getImports expr = concatMap getImports $ Lens.toListOf Core.subExpressions expr
+
+fileAsText :: File -> Text
+fileAsText File{..} = foldr (\d acc -> acc <> "/" <> d) "" (Core.components directory)
+    <> "/" <> file
+
+-- | Given an 'Import', render the contents in an HTML element that will allow
+--   users to jump to another file or domain
+renderImport :: Import -> Text -> Html ()
```

Comment:
What is the `Text` argument for? An example would be helpful here.

Reply:
I added a `doctest` there